cloud/db_cloud_test.cc

```diff
@@ -1800,6 +1809,103 @@ TEST_F(CloudTest, FileCacheOnDemand) {
+  EXPECT_EQ(basename(cloud_manifest_file), "CLOUDMANIFEST-000001");
+}
+
+// Verify that live sst files are the same after applying cloud manifest delta
+TEST_F(CloudTest, LiveFilesConsistentAfterApplyLocalCloudManifestDeltaTest) {
+  cloud_env_options_.cookie_on_open = "1";
+  OpenDB();
+  auto cenv_impl = GetCloudEnvImpl();
+  // verify that live files are the same between CLOUD_MANIFEST-cookie1 and
+  // CLOUD_MANIFEST-cookie2
+  auto verifyLiveFiles = [cenv_impl](std::string cookie1, std::string cookie2) {
+    std::vector<std::string> live_sst_files1, live_sst_files2;
+    std::string manifest_file1, manifest_file2;
+    ASSERT_OK(cenv_impl->FindAllLiveFilesWithCookie(
+        cenv_impl->GetSrcBucketName(), cenv_impl->GetSrcObjectPath(), cookie1,
+        &live_sst_files1, &manifest_file1));
+    ASSERT_OK(cenv_impl->FindAllLiveFilesWithCookie(
+        cenv_impl->GetSrcBucketName(), cenv_impl->GetSrcObjectPath(), cookie2,
+        &live_sst_files2, &manifest_file2));
+    EXPECT_NE(manifest_file1, manifest_file2);
+    EXPECT_EQ(live_sst_files1, live_sst_files2);
+  };
+
+  ASSERT_OK(db_->Put(WriteOptions(), "Hello", "World"));
+  ASSERT_OK(db_->Flush(FlushOptions()));
+
+  std::string new_cookie = "2";
+  std::string new_epoch = "dca7f3e19212c4b3";  // new epoch for MANIFEST file
+  ASSERT_OK(GetCloudEnvImpl()->ApplyLocalCloudManifestDelta(
+      dbname_, new_cookie,
+      CloudManifestDelta{GetDBImpl()->TEST_Current_Next_FileNo(), new_epoch}));
+  // check that a new manifest files are generated
+  EXPECT_OK(base_env_->FileExists(MakeCloudManifestFile(dbname_, new_cookie)));
+  EXPECT_OK(base_env_->FileExists(ManifestFileWithEpoch(dbname_, new_epoch)));
+
+  // Upload CLOUD_MANIFEST_new_cookie and MANIFEST_new_epoch to s3
+  ASSERT_OK(GetCloudEnvImpl()->UploadLocalCloudManifest(dbname_, new_cookie));
+
+  verifyLiveFiles("1" /* old cookie */, "2" /* new cookie */);
+}
+
+// After calling `ApplyLocalCloudManifestDelta`, writes should be persisted in
+// sst files only visible in new Manifest
+TEST_F(CloudTest, WriteAfterUpdateCloudManifestArePersistedInNewEpoch) {
+  cloud_env_options_.cookie_on_open = "1";
+  OpenDB();
+  ASSERT_OK(db_->Put(WriteOptions(), "Hello", "world"));
+  ASSERT_OK(db_->Flush(FlushOptions()));
+
+  std::string new_cookie = "2";
+  std::string new_epoch = "dca7f3e19212c4b3";
+  ASSERT_OK(GetCloudEnvImpl()->ApplyLocalCloudManifestDelta(
+      dbname_, new_cookie,
+      CloudManifestDelta{GetDBImpl()->TEST_Current_Next_FileNo(), new_epoch}));
+
+  // following writes are not visible for old cookie
+  ASSERT_OK(db_->Put(WriteOptions(), "Hello", "new_world"));
+  ASSERT_OK(db_->Flush(FlushOptions()));
+
+  // reopen with cookie = 1, new updates after rolling are not visible
+  CloseDB();
+  cloud_env_options_.cookie_on_open = "1";
+  OpenDB();
+  std::string value;
+  ASSERT_OK(db_->Get(ReadOptions(), "Hello", &value));
+  ASSERT_EQ(value, "world");
```

Comment:
Could you also verify that writes are visible with the new cookie? I think that's the scenario that doesn't work in the current version of the code.